index.js

```diff
@@ -693,40 +795,253 @@
     });
   };
 
-  //  testNested :: (Type, Any, Integer) -> Result
-  var testNested = function(t, x, position) {
-    var _ = '_' + String(position);
-    var $ = '$' + String(position);
-    for (var idx = 0, xs = t[_](x); idx < xs.length; idx += 1) {
-      var result = test(t[$], xs[idx]);
-      if (!result.valid) {
-        result.typePath.unshift(t);
-        result.propPath.unshift($);
-        return result;
+  //  _determineActualTypes :: (Boolean, [Type], [Object], [Any]) -> [Type]
+  var _determineActualTypes = function recur(loose, env, seen, values) {
+    if (isEmpty(values)) return [Unknown];
+    //  typeses :: [[Type]]
+    var typeses = map(values, function(value) {
+      var seen$;
+      if (typeof value === 'object' && value != null ||
+          typeof value === 'function') {
+        //  Abort if a circular reference is encountered; add the current
+        //  object to the list of seen objects otherwise.
+        if (seen.indexOf(value) >= 0) return [];
+        seen$ = seen.concat([value]);
```

Comment:
Do we really have to clone the array every time here? This function is going to be invoked a lot, it might make sense to mutate.